### ISSUE TITLE
Radio buttons show three options where there are only two.

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#3881]
I tried to create a field for the user to choose one of two options but I am seeing three, like this:

![Screenshot_2019-06-22 Edit Event John Connolly in conversation more about books](https://user-images.githubusercontent.com/10559274/59962765-84518c00-94e1-11e9-8917-f8232a4f1875.png)

I do not want the 'N/A' option and cannot see how to suppress it without recourse to CSS.

**Comment by drupol** [comment]
Set the field to "Required" and the N/A will go away.

Closing the issue, feel free to reopen if needed.

**Comment by Graham-72** [comment]
@drupol Thank you for the answer.

But I think this is a problem with the UI and needs to be fixed. It is not clear in the context what 'N/A' means and anything that causes a user significant time to try and understand needs to be improved. If instead the third radio button was labelled 'No default value' instead of 'N/A' I would probably have understood why it existed.

**Comment by klonos** [comment]
> But I think this is a problem with the UI and needs to be fixed. ...anything that causes a user significant time to try and understand needs to be improved.

Agreed. Lets wait for some more feedback with regards to possible solutions to this.

**Comment by klonos** [comment]
...related a bug discovered while testing this out: #3884

**Comment by indigoxela** [comment]
To my understanding 'N/A' is meant to be a "one-size-fits-all" string to any list fields provided as radios.
I agree, a more meaningful string could be a UI improvement.

Currently it's probably not possible to use a more meaningful string though, at least to my opinion.
Field contexts might need something like "not applicable", "no default value", "unknown", "none", "empty", "any", "not sure"...

A possible solution could be a configurable (and translatable) label text for value "_none" on a per-field basis. Not sure, if this makes sense, though, or if it would blow up UI elsewhere.

**Comment by oadaeh** [comment]
"N/A" means "not applicable", not "one-size-fits-all".

Pol is correct in that if you want it to go away, you need to specify that the field is required.

I think is anything is to be done about it, simply not adding that undesired option at all would be best. Drupal has added it for as long as I can remember, and it has always annoyed me that the system made that judgement call on my behalf without me being able to do anything about it (other than hack core).

**Comment by oadaeh** [comment]
I updated the title, because when I first read it, it seemed to me that it was meant to be [Do] Not clear...

**Comment by klonos** [comment]
> ...simply not adding that undesired option at all would be best. 

Care to elaborate please @oadaeh??

**Comment by olafgrabienski** [comment]
> simply not adding that undesired option at all would be best

AFAIK, the "N/A" option is needed for de-selecting a radio button: An editor selects a value, changes his/her mind, and wants to de-select the value (without selecting one of the other 'real' values). That wouldn't be possible without the "N/A" button.

That said, I find the naming horrible, and I like @indigoxela's idea of a configurable and translatable label text on a per-field basis.

**Comment by stpaultim** [comment]
Isn't it a little strange that if you select a dropdown widget the option is "none" and if you select radio buttons, the option is "N/A." I like "None" better. Can't we make it use that option for both widgets?

This is a quirk that I've gotten used to in Drupal and I understand why it is an issue. However, I also still remember how frustrated I was the first time I ran into this problem and how long it took me to figure out that I had to require the field to make that go away.

**Comment by klonos** [comment]
> This is a quirk that I've gotten used to in Drupal and I understand why it is an issue. However, I also still remember how frustrated I was the first time I ran into this problem and how long it took me to figure out that I had to require the field to make that go away.

Yup, this pretty much sums the UX issue up; and that's what we want to fix here.

> An editor selects a value, changes his/her mind, and wants to de-select the value...

Yes, with checkboxes things are easy, because you can check/uncheck them; but with radio sets it gets weird. The set starts w/o any option selected, but once the user selects something, they can't go back to selecting nothing. I found this interesting: https://stackoverflow.com/questions/4957207/how-to-check-uncheck-radio-button-on-click

...but we'd be overriding default browser behavior, which we can't expect users to know about. We'd be introducing a "Backdropism", so hmmm 🤔

**Comment by klonos** [comment]
...it does seem to be a pretty frequent UX issue. There's heaps of articles with workarounds:

- https://help.surveygizmo.com/help/deselect-a-radio-button
- https://micahjon.com/2013/how-to-deselect-radio-buttons-on-web-forms
- https://ux.stackexchange.com/questions/13511/why-is-it-impossible-to-deselect-html-radio-inputs
- https://coderwall.com/p/sxdvhg/jquery-radio-buttonset-allow-deselect
- https://forums.adobe.com/thread/2258949
- https://codepen.io/gessicadanielly/pen/BzadXN
- https://community.appian.com/discussions/f/user-interface/12796/how-to-deselect-a-radio-button-option

**Comment by indigoxela** [comment]
Radio combos aren't very user friendly in general, I think we agree on that. But they are sometimes necessary.

The "N/A" checkbox is technically required for un-selecting, if the field isn't required, I think this should be clear. Skipping the "N/A" radio and un-selecting via javascript doesn't seem like a good idea to me, because javascript is required then to make it work.

The generic label text could be improved. Probably "N/A" was a compromise to not add to much meaning to it.
But finding a better generic string ("one-size-fits-all" (irony!)) won't be easy, because depending on the context the actual meaning of "no choice" would require different wording.

For the field example provided by @Graham-72 it should be something like "combo image", for instance.

BTW: The workaround I'm using for use cases like that, is to make the field required and add the radio for "empty choice" as the first option (and selected) - then the label is configurable.

A cleaner solution would be to make the label text for the empty value actually configurable in field settings.

**Comment by klonos** [comment]
> The workaround I'm using for use cases like that, is to make the field required and add the radio for "empty choice" as the first option (and selected) - then the label is configurable.

Noise 👍 

> ...un-selecting via javascript doesn't seem like a good idea ...finding a better generic string won't be easy ...A cleaner solution would be to make the label text for the empty value actually configurable in field settings.

Yup. Agreed that this would provide the ideal UX for both site-builders configuring these, as well as end users using the widget.

**Comment by Graham-72** [comment]
In the particular use that I showed above as an example I could have used a single checkbox because what I needed was to allow the user a choice between two possibilities. I had used a boolean value for the field. So it came as a surprise that BD was suggesting three possibilities i.e. three buttons. I used radio buttons because, in my view, it made it very clear to the user what the choice is. And of course, with hindsight, I should have realised that making the field required was the answer to the odd result. Maybe I will remember next time.

Perhaps a simple way forward might be to label the extra radio button 'clear' or 'clear buttons' because that is what it does, doesn't it? This surely would be a lot more helpful than the ambiguous 'N/A'. Though I'm not sure what this implies if the field is boolean.

**Comment by stpaultim** [comment]
> Isn't it a little strange that if you select a dropdown widget the option is "none" and if you select radio buttons, the option is "N/A." I like "None" better. Can't we make it use that option for both widgets?

Is there a reason why the dropdown widget says "none" and the radio button says "N/A." Wouldn't using "none" for both be better than the current situation?

**Comment by oadaeh** [comment]
@stpaultim it has been that was since it was first implemented in Drupal in commit 607e9626d5af265b18e8319b156bb0fda3445cd4 back on Tue Feb 3 17:30:13 2009 +0000.
